core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java

```diff
@@ -0,0 +1,183 @@
+import static java.util.logging.Level.FINER;
+import static java.util.logging.Level.SEVERE;
+import static java.util.logging.Level.WARNING;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.logging.Logger;
+
+import hudson.Extension;
+import hudson.model.AsyncPeriodicWork;
+import hudson.model.Item;
+import hudson.model.Job;
+import hudson.model.LogRotatorConfiguration;
+import hudson.model.LogRotatorMapping;
+import hudson.model.TaskListener;
+import jenkins.model.BuildDiscarder;
+import jenkins.model.GlobalConfiguration;
+import jenkins.model.Jenkins;
+
+/**
+ * Periodic task that checks if it's time to rotate build logs.
+ * 
+ * @author awitt
+ * @since 2.203
+ *
+ */
+@Extension
+public class LogRotatorPeriodicTask extends AsyncPeriodicWork {
+    
+    private static final Logger LOGGER = Logger.getLogger( LogRotatorPeriodicTask.class.getName() );
+    
+    public LogRotatorPeriodicTask() {
+        super( Messages.LogRotatorPeriodicTask_title() );
+    }
+
+    @Override
+    protected void execute(TaskListener listener) throws IOException, InterruptedException {
+        
+        LogRotatorConfiguration config = GlobalConfiguration.all().get( LogRotatorConfiguration.class );
+        
+        if( config != null ) {
+            // there is config
+            if( config.isEnableRotation() && config.getUpdateIntervalHours() > 0 ) {
+                // and we should update periodically
+                
+                long millis_since_last_update = System.currentTimeMillis() - config.getLastRotated();
+                double hours_since_last_update = millis_since_last_update / 1000.0 / 60.0 / 60.0;
+                
+                if( hours_since_last_update >= config.getUpdateIntervalHours() ) {
+                    // actually do it
+                    LOGGER.log( getNormalLoggingLevel(), Messages.LogRotatorPeriodicTask_starting() );
+                    
+                    Jenkins jenkins = Jenkins.getInstanceOrNull();
+                    
+                    if( null == jenkins ) { return; }
+                    
+                    for( Item item : jenkins.getAllItems() ) {
+                        
+                        for( Job<?,?> job : item.getAllJobs() ) {
+                                                        
+                            BuildDiscarder discarder = job.getBuildDiscarder();
+                                                        
+                            try {
+                            
+                                if( discarder instanceof LogRotator ) {
+                                    // the job has defined its own LogRotator                                    
+                                    switch( config.getPolicyForJobsWithCustomLogRotator() ) {
+                                    case NONE:
+                                        LOGGER.log( FINER, Messages.LogRotatorPeriodicTask_hasRotator_policyNONE( job.getFullName() ) );
```

Comment:
Why? I haven't seen this before, and it seems to work without it.

Reply:
Oh! It's to defer processing of the message until it actually has to get rendered in a log, speeding up the pass through the actual code, isn't it?

Reply:
Yes, for any processing to build your logging string this defers its creation till its needed. In the past a user would have to construct the whole string before calling LOGGER.log but this created the pattern where users would check outside the function leading to
```
if (LOGGER.isLoggable(FINER)) {
LOGGER.log()
}
```
Using a lambda just simplifies it since internally `log` checks if the logger is loggable at the specified level first.